manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,654 @@
+        } catch (Exception e) {
+            //If the payload wasn't parsed, then it means that it is either a response to a command,
+            //or it's genuinely a wrong payload.
+
+            mapper = new ObjectMapper();
+            TeltonikaResponsePayload response = mapper.readValue(payloadContent, TeltonikaResponsePayload.class);
+            AttributeMap map = new AttributeMap();
+            // TODO: Do we want to remove the value from TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME?
+            map.addAll(new Attribute<>
+                    (
+                            new AttributeDescriptor<>(TELTONIKA_DEVICE_RECEIVE_COMMAND_ATTRIBUTE_NAME, ValueType.TEXT),
+                            response.rsp
+                    )
+            );
+            return map;
+        }
+
+    }
+
+
+    /**
+     * Returns an {@code Optional<Attribute<AssetStateDuration>>}, that {@code ifPresent()}, represents
+     * the Duration for which the predicate returned true.
+     *
+     * @param previousValue The old attribute state (Or the latest datapoint that exists)
+     * @param newValue The new attribute value
+     * @param pred A Predicate that describes the state change
+     * @param ref An AttributeRef that describes which asset and attribute this pertains to.
+     * @return An Optional Attribute of type AssetStateDuration that represents the Duration for which the predicate returned true.
+     */
+    //TODO: Change this to only use an AttributeRef and a Predicate.
+    private Optional<Attribute<?>> assetChangedTripState(Attribute<?> previousValue, Attribute<?> newValue, AttributePredicate pred, AttributeRef ref) {
+        //We will first check if the predicate fails for the new value, and then check if the predicate is true for the previous value.
+        //In that way, we know that the state change happened between the new and previous values.
+
+        if (newValue.getValue().isEmpty()) {
+            getLogger().warning("The new value is empty!");
+            return Optional.empty();
+        } else if (previousValue.getValue().isEmpty()) {
+            getLogger().warning("The old value is empty!");
+            return Optional.empty();
+        }
+
+        //TODO: Understand what happens with negate(), does it change states to the object itself?
+        boolean newValueTest =      pred.value.asPredicate(timerService::getCurrentTimeMillis).test(newValue        .getValue().get());
+        boolean previousValueTest = pred.value.asPredicate(timerService::getCurrentTimeMillis).test(previousValue   .getValue().get());
+        //If the predicate fails, then no changes need to happen.
+
+        // newValue is not 1, previousValue == 1
+        if(!(!newValueTest && previousValueTest)) {
+            return Optional.empty();
+        }
+
+        // Grab all datapoints (To be replaced by AssetDatapointValueQuery)
+        List<AssetDatapoint> valueDatapoints = AssetDatapointService.getDatapoints(ref);
```

Comment:
It would be more efficient to keep in memory the attributes that have an active session rather than searching back through historical data on each publish

Reply:
Would this be stored inside some sort of instance variable in TeltonikaDevice? Hence we could easily link the incoming data/state change to the device producing them. Thanks for the idea!

Reply:
But would it be okay if the state was deleted, in case OpenRemote is shut down? If it was stored in the database, we could simply check the last time the value was equal to the current value.